common/finalizer/finalizer.go

```diff
@@ -0,0 +1,169 @@
+	mu        sync.Mutex  // lock protects access to callbacks map
+	finalized atomic.Bool // indicating if finalizer has been executed
+	callbacks map[string]func(context.Context) error
+}
+
+func New(
+	logger log.Logger,
+) *Finalizer {
+	return &Finalizer{
+		logger:    logger,
+		callbacks: make(map[string]func(context.Context) error),
+	}
+}
+
+// Register adds a callback to the finalizer.
+// Returns an error if the ID is already registered, or when the finalizer already ran.
+func (f *Finalizer) Register(
+	id string,
+	callback func(context.Context) error,
+) error {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; ok {
+		return FinalizerDuplicateIdErr
+	}
+	f.callbacks[id] = callback
+	return nil
+}
+
+// Deregister removes a callback from the finalizer.
+// Returns an error if the ID is not found, or when the finalizer already ran.
+func (f *Finalizer) Deregister(
+	id string,
+) (err error) {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; !ok {
+		return FinalizerUnknownIdErr
+	}
+	delete(f.callbacks, id)
+	return nil
+}
+
+// Run executes all registered callback functions within the given timeout (zero timeout skips execution).
+// It can only be invoked once; calling it again has no effect.
+// Returns the number of completed callbacks.
+func (f *Finalizer) Run(
+	pool *goro.AdaptivePool,
+	timeout time.Duration,
+) int {
+	if !f.finalized.CompareAndSwap(false, true) {
+		f.logger.Warn("finalizer skipped: called more than once")
+		return 0
+	}
+
+	if timeout == 0 {
+		f.logger.Warn("finalizer skipped: zero timeout")
+		return 0
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	totalCount := len(f.callbacks)
+	f.logger.Info("finalizer starting",
+		tag.NewInt("items", totalCount),
+		tag.NewDurationTag("timeout", timeout))
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	done := make(chan struct{})
+	completionChannel := make(chan struct{})
+	go func() {
+		for _, callback := range f.callbacks {
+			// NOTE: Once `pool.Stop` is called due to a timeout, any remaining calls to `pool.Do` will do nothing.
+			pool.Do(func() {
+				defer func() { completionChannel <- struct{}{} }()
+				_ = callback(ctx)
+			})
+		}
+		done <- struct{}{}
+	}()
+
+	var completedCallbacks int
+	for {
+		select {
+		case <-completionChannel:
+			completedCallbacks += 1
+
+			if completedCallbacks == totalCount {
+				f.logger.Info("finalizer completed",
+					tag.NewInt("completed", completedCallbacks))
+				return completedCallbacks
+			}
+
+		case <-ctx.Done():
+			pool.Stop()
+
+			f.logger.Error("finalizer timed out",
+				tag.NewInt("completed", completedCallbacks),
+				tag.NewInt("unfinished", totalCount-completedCallbacks))
+
+			// ensure that goroutine completed and does not access callbacks anymore before releasing the lock
+			<-done
```

Comment:
nit: I'd probably always do this maybe in a defer or something. I understand in completed case it will always be unblocked, but that took me some time to realize that when reading the code. No strong opinion.

Reply:
Good feedback; I've added a defer!